Add an OAuth2 extension (#3078)
quarkusio/quarkus: docs/src/main/asciidoc/oauth2-guide.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - Using OAuth2 RBAC

include::./attributes.adoc[]
:extension-name: Elytron Security OAuth2

> :oauth2-token: OAuth2 tokens

Reviewer: Hmmm, what's the purpose of this one?

Author: I create this guide based on the JWT one and kept the variables in place, I can remove them if you want.

This guide explains how your Quarkus application can utilize {oauth2-token} to provide secured access to the JAX-RS endpoints.

> If your OAuth2 Authentication server provides JWT tokens, you should use link:jwt-guide.html[MicroProfile JWT RBAC] instead, this extension aims to be used with obscure tokens and validate the token by calling an introspection endpoint.

Reviewer: `obscure`?

Author: `obscure` tokens are the opposite of meaningfull tokens like JWT ones as they didn't carry any informations. This terms comes from the discussion on this PR, maybe I can replace it as it's not a commonly used one but I don't know what to use ?

Reviewer: `exotic` maybe if it's the meaning you want to convey?

Reviewer 2: Isn't the correct term `opaque token` ?

Author: @sebastienblanc yes, `opaque` is the correct term, I will update the documentation with this.

Author: done

== Configuration

> 

Reviewer: Maybe a small intro would be useful.

Author: Maybe something like this ?

```
OAuth 2 is an authorization framework that enables applications to obtain access to an HTTP resource on behalf of a user. 
It can be used to implement an application authentication mechanism based on tokens by delegating to an external server (the authentication server) the user authentification and providing a token for the authentication context.
```

In fact it's always complex to define what is OAuth2 because it's not an authentication protocole but it can be used to build one ...

Reviewer: Your proposal looks fine to me except for `s/OAuth 2/OAuth2/`.

Author: done

[cols="<m,<m,<2",options="header"]
|===
|Property Name|Default|Description

> |quarkus.oauth2.enabled|true|Determine if the OAuth2 extension is enabled. Enabled by default if you include the elytron-security-oauth2 dependency, so this would be used to disable it.

Reviewer: backticks around the dependency name would be nice.

Maybe also around the properties, depending on how it is done in the other guides.

Author: done.

backticks around properties when use a table is not the common usage on the other guides.

|quarkus.oauth2.role-claim|scope|The claim that is used in the introspection endpoint response to load the roles
|===

== Solution

We recommend that you follow the instructions in the next sections and create the application step by step.
However, you can skip right to the completed example.

Clone the Git repository: `git clone {quickstarts-clone-url}`, or download an {quickstarts-archive-url}[archive].

The solution is located in the `using-oauth2-rbac` {quickstarts-archive-url}[directory].

== Creating the Maven project

First, we need a new project. Create a new project with the following command:

[source,shell,subs=attributes+]
----
mvn io.quarkus:quarkus-maven-plugin:{quarkus-version}:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=using-oauth2-rbac \
    -DclassName="org.acme.oauth2.TokenSecuredResource" \
    -Dpath="/secured" \
    -Dextensions="resteasy-jsonb, security-oauth2"
----

This command generates the Maven project with a REST endpoint and imports the `elytron-security-oauth2` extension, which includes the {auth2-token} support.

> If you don't want to use the maven plugin, you can just include the dependency in your pom.xml:

Reviewer: ```suggestion
If you don't want to use the Maven plugin, you can just include the dependency in your pom.xml:
```

Author: done

[INFO] --------------------------------[ jar ]---------------------------------
...
[INFO] --- quarkus-maven-plugin:999-SNAPSHOT:dev (default-cli) @ using-oauth2-rbac ---
Listening for transport dt_socket at address: 5005
2019-07-16 09:58:09,753 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-07-16 09:58:10,884 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 1131ms
2019-07-16 09:58:11,385 INFO  [io.quarkus] (main) Quarkus 0.20.0 started in 1.813s. Listening on: http://[::]:8080
2019-07-16 09:58:11,391 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy, resteasy-jsonb, security, security-oauth2]

----

Now that the REST endpoint is running, we can access it using a command line tool like curl:

.curl command for /secured/permit-all
[source,shell]
----
$ curl http://127.0.0.1:8080/secured/permit-all; echo
hello + anonymous, isSecure: false, authScheme: null
----

We have not provided any token in our request, so we would not expect that there is any security state seen by the endpoint, and the response is consistent with that:

* user name is anonymous
* isSecure is false as https is not used

> * authScheme is null

Reviewer: Some backticks around isSecure and authScheme would be nice.

import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.SecurityContext;

@Path("/secured")
@ApplicationScoped
public class TokenSecuredResource {

    @GET()
    @Path("permit-all")
    @PermitAll
    @Produces(MediaType.TEXT_PLAIN)
    public String hello(@Context SecurityContext ctx) {
        Principal caller =  ctx.getUserPrincipal();
        String name = caller == null ? "anonymous" : caller.getName();
        String helloReply = String.format("hello + %s, isSecure: %s, authScheme: %s", name, ctx.isSecure(), ctx.getAuthenticationScheme());
        return helloReply;
    }

    @GET()
    @Path("roles-allowed") // <1>
    @RolesAllowed({"Echoer", "Subscriber"}) // <2>
    @Produces(MediaType.TEXT_PLAIN)
    public String helloRolesAllowed(@Context SecurityContext ctx) {
        Principal caller =  ctx.getUserPrincipal();
        String name = caller == null ? "anonymous" : caller.getName();
        String helloReply = String.format("hello + %s, isSecure: %s, authScheme: %s", name, ctx.isSecure(), ctx.getAuthenticationScheme());
        return helloReply;
    }
}
----

> <1> This new endpoint will be located at /secured/roles-allowed

Reviewer: backticks around the URL

Author: done

> <2> @RolesAllowed indicates that the given endpoint is accessible by a caller if they have either a "Echoer" or "Subscriber" role assigned.

Reviewer: backticks for the annotation

Author: done

After you make this addition to your `TokenSecuredResource`, try `curl -v http://127.0.0.1:8080/secured/roles-allowed; echo` to attempt to access the new endpoint. Your output should be:

.curl command for /secured/roles-allowed
[source,shell]
----
$ curl -v http://127.0.0.1:8080/secured/roles-allowed; echo
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> GET /secured/roles-allowed HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 401 Unauthorized
< Connection: keep-alive
< Content-Type: text/html;charset=UTF-8
< Content-Length: 14
< Date: Sun, 03 Mar 2019 16:32:34 GMT
<
* Connection #0 to host 127.0.0.1 left intact
Not authorized
----

> Excellent, we have not provided any Oauth2 token in the request, so we should not be able to access the endpoint, and we were not. Instead we received an HTTP 401 Unauthorized error. We need to obtain and pass in a valid OAuth2 token to access that endpoint. There are two steps to this, 1) configuring our {extension-name} extension with information on how to validate the token, and 2) generating a matching token with the appropriate claims.

Reviewer: ```suggestion
Excellent, we have not provided any OAuth2 token in the request, so we should not be able to access the endpoint, and we were not. Instead we received an HTTP 401 Unauthorized error. We need to obtain and pass in a valid OAuth2 token to access that endpoint. There are two steps to this, 1) configuring our {extension-name} extension with information on how to validate the token, and 2) generating a matching token with the appropriate claims.
```

Author: done

== Configuring the {extension-name} Extension Security Information

> In the <<Configuration>> section we introduce the `application.properties` file that affect the {extension-name} extension.

Reviewer: ```suggestion
In the <<Configuration>> section we introduce the configuration properties that affect the {extension-name} extension.
```

=== Setting up application.properties

>  For part A of step 1, create a using-oauth2-rbac/src/main/resources/application.properties with the following content:

Reviewer: backticks

Author: done

.application.properties for TokenSecuredResource
[source, properties]
----
quarkus.oauth2.client-id=client_id
quarkus.oauth2.client-secret=secret
quarkus.oauth2.introspection-url=http://oauth-server/introspect
----

> You need to specify the introspection URL of your authentication server and the client_id/client_secret that your application will use to authenticate itself to the authentication server.

Reviewer: `client-id`, `client-secret` with dashes and backticks

Author: done

> The extension will then use these informations to validate the token and recover the information associate with it.

Reviewer: ```suggestion
The extension will then use this information to validate the token and recover the information associated with it.
```

Author: done

=== Generating a token

> You need to obtains the token from a standard OAuth2 authentication server (https://www.keycloak.org/[Keycloak] for example) using the token endpoint.

Reviewer 3: .You need to obtain....

Author: done

> You can find bellow a curl example of such call for a client_credential flow:

Reviewer: ```suggestion
You can find below a curl example of such call for a `client_credentials` flow:
```

Author: done

[source,shell]
----
curl -X POST "http://oauth-server/token?grant_type=client_credentials" \
-H  "Accept: application/json" -H  "Authorization: Basic Y2xpZW50X2lkOmNsaWVudF9zZWNyZXQ="
----

> It should respond something like that

Reviewer 3: It should respond with something like that...

Author: done

[source,json]
----
{"access_token":"60acf56d-9daf-49ba-b3be-7a423d9c7288","token_type":"bearer","expires_in":1799,"scope":"READER"}
----

> == Finally, Secured Access to /secured/roles-allowed

Reviewer 3: Finally, we have a secured access to `/secured/roles-allowed`. Now ....

Reviewer: It's a title, not a sentence. I think it's correct, unless I misinterpret your comment.

Author: It's a title, not sure I can use backticks in title

Reviewer 3: Sure, Ignore comment.

> Now let's use this to make a secured request to the /secured/roles-allowed endpoint

Reviewer: ```suggestion
Now let's use this to make a secured request to the `/secured/roles-allowed` endpoint
```

Author: done

.curl Command for /secured/roles-allowed With a token
[source,shell]
----
$ curl -H "Authorization: Bearer 60acf56d-9daf-49ba-b3be-7a423d9c7288" http://127.0.0.1:8080/secured/roles-allowed; echo
hello + client_id isSecure: false, authScheme: OAuth2
----

Success! We now have:

* a non-anonymous caller name of client_id
* an authentication scheme of OAuth2

== Roles mapping

> Roles are mapped from one of the claims of the introspection endpoint response. By default, it's the `scope` claim. Roles are obtains by splitting the claim with a space separator. If the claim is an array, no splitting are done, the roles are obtains from the array.

Reviewer: ```suggestion
Roles are mapped from one of the claims of the introspection endpoint response. By default, it's the `scope` claim. Roles are obtained by splitting the claim with a space separator. If the claim is an array, no splitting is done, the roles are obtained from the array.
```

Author: done

[INFO] Scanning for projects...
...
[using-oauth2-rbac-runner:25602]     universe:     493.17 ms
[using-oauth2-rbac-runner:25602]      (parse):     660.41 ms
[using-oauth2-rbac-runner:25602]     (inline):   1,431.10 ms
[using-oauth2-rbac-runner:25602]    (compile):   7,301.78 ms
[using-oauth2-rbac-runner:25602]      compile:  10,542.16 ms
[using-oauth2-rbac-runner:25602]        image:   2,797.62 ms
[using-oauth2-rbac-runner:25602]        write:     988.24 ms
[using-oauth2-rbac-runner:25602]      [total]:  43,778.16 ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  51.500 s
[INFO] Finished at: 2019-06-28T14:30:56-07:00
[INFO] ------------------------------------------------------------------------

$ ./target/using-oauth2-rbac-runner
2019-03-28 14:31:37,315 INFO  [io.quarkus] (main) Quarkus 0.20.0 started in 0.006s. Listening on: http://[::]:8080
2019-03-28 14:31:37,316 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy, resteasy-jsonb, security, security-oauth2]
----

== Explore the Solution

> The solution repository located in the `using-oauth2-rbac` {quickstarts-archive-url}[directory] contains all of the versions we have worked through in this quickstart guide. 

Reviewer: There's a good chance it won't be true (the fact that you have all the versions?).

Author: rephrased
